rewrite-core/src/main/java/org/openrewrite/RecipeScheduler.java

```diff
@@ -141,14 +142,14 @@
             }
         }
 
+        final TreeVisitor<?, ExecutionContext> singleSourceApplicableTest = recipe.getSingleSourceApplicableTest();
```

Comment:
My only concern with this change is that if a singleSourceApplicableTest has mutable state, it might be an issue because multiple threads could be calling this concurrently.

Reply:
Visitors in general are assumed to be instantiated one per source file.